src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/EntrypointFactoryInterface.php

```diff
@@ -0,0 +1,27 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\SecurityBundle\DependencyInjection\Security\Factory;
+
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+
+/**
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ *
+ * @experimental in 5.1
+ */
+interface EntrypointFactoryInterface
+{
+    /**
+     * Creates the entry point and returns the service ID.
+     */
+    public function createEntryPoint(ContainerBuilder $container, string $id, array $config, ?string $defaultEntryPointId): string;
```

Comment:
I have it on my list to look at the entrypoint stuff more closely. But my gut reaction is this: let's fix them (or try to... or maybe in a PR after this) :). What I mean is this:

* It's always been confusing that your firewall may have multiple auth mechanisms that have their own "entry point"... and one is chosen seemingly at random :). I know it's not random, but why does the entrypoint from `form_login` "win" over `http_basic` if I have both defined under my firewall?

* Since we're moving to a new system, why not throw an exception the *moment* that a firewall has multiple entrypoints available to it. Then we *force* the user to choose the *one* entrypoint that should be used.

There is already an `entrypoint` config under the firewall to control this. I think this should be *required* when there are more than one entrypoint under a firewall. BUT, I believe this key currently accepts only a service id. It would need to be expanded to allow a service id OR a "authenticator factory name" - so that I could do either of these:

```yml
firewalls:
    main:
        form_login: true

        # use a service id like now
        entry_point: App\Security\MyCustomEntrypointWhichCouldBeAnAuthenticatorTooItDoesntMatter

        # OR point to one of your authenticator keys
        entry_point: form_login

```

Reply:
Agreed, but let's do it in another PR (I think we simply don't have time now to fix this, as it requires quite a bit of research/code).

Reply:
Fair enough! I hope it will be easy (collecting all of the entrypoints in `createAuthenticationListeners()` and throwing an exception if there are multiple, or choosing one if argument 5 to that method (`$defaultEntryPoint`, which is the `entry_point` from the firewall) is a key in the array of entrypoints, use that) - but we can save until later.